### ISSUE TITLE
Enhancement assessment: VK_PRIMITIVE_TOPOLOGY_TRIANGLE_FAN

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/MoltenVK#1419]
Metal does not support triangle fans (`VK_PRIMITIVE_TOPOLOGY_TRIANGLE_FAN`). This is reported accurately by MoltenVK as part of `VkPhysicalDevicePortabilitySubsetFeaturesKHR` in the `VK_KHR_portability_subset` extension.

However, some apps, including many CTS tests, continue to need triangle fans, which opens the question of whether MoltenVK should attempt to support triangle fans, regardless of Metal support.

Triangle fans can be treated as triangle strips, with a different vertex order. Specifically, a triangle fan can be converted to a triangle strip by reorganizing the vertex indexing using a standard algorithm:

```
    if( vtxIdx % 2 == 0 )
        vtxIdx = vtxIdx / 2;
    else
        vtxIdx = vtxCnt - 1 - vtxIdx / 2;
```

There are a couple of options about where and how this remapping should be done. I've done some initial research into options within MoltenVK, and present them here for discussion. There are no doubt other options, and I may be over-or-under-estimating the complexity of each option in my initial review.

#### Remap Vertices or Vertex Indices Prior to Draw

In this first option, since the vertex data can be updated in the vertex buffer anytime before a draw call is encoded, including via other GPU activity, the remapping of the vertices would need to be done via copying the vertices or vertex indices into a temporary `MTLBuffer` using a Compute Shader encoded just before the draw command.

We could add a `kMVKGraphicsStageVertexRemap` to `MVKGraphicsStage` enumeration, and include it in the results of `MVKGraphicsPipeline::getStages()` when the pipeline requires triangle fan topology. The various `MVKCmdDraw*::encode()` functions could then incorporate this into their stage management operations, similarly to how they incorporate Compute Shaders as part of tessellation operations.

The scheduling of the `kMVKGraphicsStageVertexRemap` would need to take into consideration both tessellated and non-tessellated pipeline handling, and the passing of the temporary vertex buffer in place of the original buffer. This could be quite a complex modification, given the complexity of the existing `MVKCmdDraw*::encode()` functions.

#### Remap Each Vertex in the `vertex` Shader

In this second option, the MSL of the `vertex` shader could be modified to perform the above remapping on each vertex, by modifying the effective value of `glVertexID` before grabbing the vertex data from the array of vertices.

Currently, both MoltenVK and SPIRV-Cross support directly accessing the vertex as an `[[ attribute(n) ]]` in the `[[ stage_in ]]` vertex attribute structure, and do not support random indexing into the vertex array. For this remapping option to be available, both MoltenVK and SPIRV-Cross would require significant modifications to their handling of vertex attributes. This is made more complex by certain vertex attribute modifications within MoltenVK to work around differences in vertex attribute offset and stride handling between Vulkan and MoltenVK.

@cdavis5e @HansKristian-Work I would appreciate any thoughts from you.

**Comment by mlfarrell** [comment]
Hey guys.  any update on this one?  Is triangle support coming for moltenvk?

**Comment by mlfarrell** [comment]
![Screenshot 2023-05-19 at 8 55 49 AM](https://github.com/KhronosGroup/MoltenVK/assets/7664370/4368b345-e38d-4232-aabc-a480256acdb8)

**Comment by jeroenbakker-atmind** [comment]
Hi Blender developer responsible for the Vulkan port. 

Although our use case is compatibility/maintainability. Blender already has support for OpenGL and Metal. Our user base can have create scripts using Python that will draw overlays onto any editor. There are massive amounts of these scripts and out of our hand. And it isn't realistic to drop this feature.

Our current Metal Backend already has a work-around for triangle fans. Although we could port this one over to Vulkan and only activate it when MoltenVK reports that it isn't available; this could become a maintenance issue in the longer run as this code-path might not be tested. Regular developers and all users would use Metal, gpu developers would want to use Apple hardware as it is the only platform that can run and perform compile time validation on all supported 3 backends (OpenGL/Vulkan/Metal).

We don't think it is a show stopper, but it could be that we would not support MoltenVK on Apple devices as more work-arounds are needed to get the same support level. Adding this to MoltenVK would ease cross platform support.

On the other hand, what would be the effect when 'all applications' consider to remove using triangle fans. Vulkan drivers/sillicon would still need to support it as it is part of the Vulkan spec and we might have a soft-deprecated part of the API.

**Comment by billhollings** [comment]
@jeroenbakker-atmind 

Thanks for your recommendation, and for the background on Blender's interest in this.

Don't bother adding a special path in your Vulkan-MoltenVK backend. Triangle-fans are now fairly high on our priority list, and we will be adding support fairly soon. 

Since you stress consistency, if you care to share how Blender handles it in your Metal backend, we can strive to be consistent with that functionality.

BTW...we were unaware that Blender supports MoltenVK as part of the backend options. Do you bundle MoltenVK with Blender on macOS, or do you rely on the Vulkan SDK to manage installing and finding MoltenVK, as it finds any Vulkan driver on any other platform?

**Comment by jeroenbakker-atmind** [comment]
That is good news.

Quick overview of Blender GPU backend architecture: Due our long existence and different versions of GPU backends we have developed our own GPU abstraction layer called gpu module (https://projects.blender.org/blender/blender/src/branch/main/source/blender/gpu). All GPU commands pass this module one way or another. This module is also accessible to add-on developers. The GPU has different backends for OpenGL/Metal/Vulkan. At the start of Blender one backend is loaded. The active backend can be chosen in the preferences. As our origin is OpenGL some OpenGL features had to be implemented on both Metal/Vulkan for example we track state a globally for bindings and they are bound to the shader just before the drawing/dispatch command is sent to the active backend.

In our backend the topology primitive is part of the index buffer data. When the metal backend detects unsupported configurations it converts the index buffer to a new GPU buffer https://projects.blender.org/blender/blender/src/branch/main/source/blender/gpu/metal/mtl_index_buffer.mm#L307. As input data is CPU bound it can happen on the fly when uploading it to GPU bound memory.

The triangle fan is just the first issue we detected when starting Blender using MoltenVK. Fun fact: The initial development of the Vulkan Backend has been conducted using MoltenVK using test cases.

Our current implementation is to distribute MoltenVK as an internal library. It hasn't been decided if this is the best option and would be available to end users. Changes can happen to policy and implementation before we do an initial release (ETA is to be available for early adaptors in Blender 4.1).
